### PR TITLE
[feat] CI essential multi-product architecture changes

### PR DESCRIPTION
**Pull request by phoinixgrr**
This takes care of aligning mattermost-server CircleCI pipeline to the multi-product architecture.
Specifically the following are performed:

- seperation of mattermost-server and mattermost-webapp clone/setup stages
- introduction of clone/setup stage for focalboard
- change to using `cimg/base` image instead of  `mattermost/mattermost-build-webapp` when cloning/setup mattermost-server
- introduction of `configure-multirepo` stage, in which multi-module workspaces (golang) configuration is being performed for mattermost-server and focalboard
- introduction of parallel build of focalboard along with mattermost-server
- reorganization of step dependencies, to accomodate `configure-multirepo` step

Build: https://app.circleci.com/pipelines/github/mattermost/mattermost-server/36027/workflows/ce970493-3540-4552-bd77-1a9ae89550aa

Ticket: https://mattermost.atlassian.net/browse/CLD-3801

Signed-off-by: Akis Maziotis <akis.maziotis@mattermost.com>

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -570,48 +637,17 @@ workflows:
...
           context: mattermost-ci-pr-builds-s3
           requires:
-            - build
+            - build-mattermost-server
+            - build-focalboard
       - build-docker:
           context: matterbuild-docker
           requires:
             - upload-s3-sha
-      - test:
```
Comment (pfltdv) [inline .circleci/config.yml:582]: @phoinixgrr  is it possible to remove only focalboard package tests from the pipeline. We will still need server tests.

File: .circleci/config.yml [hunk 2]
```diff
@@ -304,6 +318,38 @@ jobs:
...
+    working_directory: ~/mattermost
+    steps:
+      - attach_workspace:
+          at: ~/mattermost/
+      - run:
+          command: |
+            cd mattermost-server
+            make setup-go-work
+            cd ../focalboard
+            EXCLUDE_ENTERPRISE=TRUE make templates-archive
```
Comment (wiggin77) [inline .circleci/config.yml:345]: The `EXCLUDE_ENTERPRISE` env var is only used by the `setup-go-work` task.  It should be added there instead?

File: .circleci/config.yml (hunk 1 above)
Comment (phoinixgrr) [reply .circleci/config.yml:582]: Aha, Thought we were disabling DB Tests for Mattermost-server also. 
Makes sense. Sure, will push a commit to fix this. 

Thanks @pfltdv

**Comment by agnivade** [comment]
Left a question in the channel thread: https://community.mattermost.com/core/pl/5m41naeydtdff8sj1oswj8i7hy

**Comment by isacikgoz** [comment]
@phoinixgrr Right now we didn't run mysql/postgres tests right? I couldn't find the report for those.

File: .circleci/config.yml (hunk 1 above)
Comment (phoinixgrr) [reply .circleci/config.yml]: Tests are now enabled with https://github.com/mattermost/mattermost-server/pull/20783/commits/6979e8685a47bc2fe0708195c7cb1075f1baabe4

Tests passing 
https://app.circleci.com/pipelines/github/mattermost/mattermost-server/36047/workflows/5172c280-48c4-4e1a-8b66-d7ea31213232

**Comment by phoinixgrr** [comment]
@agnivade @isacikgoz @wiggin77  

Tests are now enabled with https://github.com/mattermost/mattermost-server/pull/20783#discussion_r941098692
Circle CI pipeline for mattermost-server is now green. 


There is the following test that fails: https://app.circleci.com/pipelines/github/mattermost/enterprise/17872/workflows/0549cc09-04b1-4982-96fa-149d5a480fc3/jobs/111795

```
unning golangci-lint
/go/src/github.com/mattermost/mattermost-server/bin/golangci-lint run ./...
WARN [runner] The linter 'golint' is deprecated (since v1.41.0) due to: The repository of the linter has been archived by the owner.  Replaced by revive. 
WARN [runner] Can't run linter goanalysis_metalinter: buildir: failed to load package cluster: could not load export data: no export data for "github.com/mattermost/enterprise/cluster" 
WARN [linters context] structcheck is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649. 
ERRO Running error: 1 error occurred:
	* can't run linter goanalysis_metalinter: buildir: failed to load package cluster: could not load export data: no export data for "github.com/mattermost/enterprise/cluster"
```

Kindly let us know of next steps

**Comment by agnivade** [comment]
This error means either the go.work file is not correct, or the enterprise repo is not cloned at the same root level as the server repo.

**Comment by phoinixgrr** [comment]
@agnivade 

As I understand, This failure is not part of the changes introduced in this PR. 
I can see that these are tests, that are running on the enterprise repo. 

Should we force-merge this and raise another ticket to troubleshoot the failures in enterprise

**Comment by phoinixgrr** [comment]
> This error means either the go.work file is not correct, or the enterprise repo is not cloned at the same root level as the server repo.

@agnivade 

These are the go.work file contests on the system that runs the aforementioned test. 
```
cat go.work
go 1.18

use ./
use ../enterprise
use ../focalboard/server
use ../focalboard/mattermost-plugin
```

This is the directory structure of the filesystem on the system that runs the aforementioned test. 
```
tree -d -L 3 /go/src/github.com
/go/src/github.com
`-- mattermost
    |-- enterprise
      [...] (obscured)
    |-- focalboard
    |   |-- build
    |   |-- docker
    |   |-- docker-testing
    |   |-- docs
    |   |-- experiments
    |   |-- import
    |   |-- linux
    |   |-- mac
    |   |-- mattermost-plugin
    |   |-- noticegen
    |   |-- server
    |   |-- webapp
    |   |-- website
    |   `-- win-wpf
    |-- mattermost-server
    |   |-- api4
    |   |-- app
    |   |-- audit
    |   |-- bin
    |   |-- build
    |   |-- cmd
    |   |-- config
    |   |-- db
    |   |-- doc
    |   |-- einterfaces
    |   |-- fonts
    |   |-- i18n
    |   |-- imports
    |   |-- jobs
    |   |-- manualtesting
    |   |-- model
    |   |-- plugin
    |   |-- product
    |   |-- scripts
    |   |-- services
    |   |-- shared
    |   |-- store
    |   |-- templates
    |   |-- testlib
    |   |-- tests
    |   |-- utils
    |   |-- web
    |   `-- wsapi
    `-- mattermost-webapp
        |-- actions
        |-- build
        |-- client
        |-- components
        |-- dist
        |-- e2e
        |-- fonts
        |-- i18n
        |-- images
        |-- packages
        |-- plugins
        |-- reducers
        |-- sass
        |-- selectors
        |-- sounds
        |-- store
        |-- stores
        |-- tests
        |-- types
        `-- utils

82 directories
```


Kindly let us know of next steps.

File: .circleci/config.yml (hunk 2 above)
Comment (wiggin77) [reply .circleci/config.yml:345]: @phoinixgrr , this doesn't make sense to me.  While I do understand how makefile prerequisites work, I don't see how adding `EXCLUDE_ENTERPRISE` to the `make templates-archive` has any effect.

In the `make setup-go-work` task two lines above, a go.work file will be created.  If `EXCLUDE_ENTERPRISE` is already defined somewhere as true then it will not include enterprise repo, otherwise it will.  Regardless, the file will be created (if not already existing).  Once it gets to the `make templates-archive` task, the go.work file already exists and won't be recreated because the `-f` flag (force) is not set, therefore the `EXCLUDE_ENTERPRISE` is ignored.

If you want the enterprise repo excluded then that var should appear prefixed to `make setup-go-work`, otherwise it should be removed altogether.

Comment (phoinixgrr) [reply .circleci/config.yml:345]: @wiggin77  

Thanks for the feedback. 

Part of the PR deliverables is to successfully parallel build focalboard and mattermost-server. 
Screenshot 2022-08-09 at 16 56 10
 
During our testing, without running the `templates-archive` target (which also runs the `setup-go-work`) the build fails.

Comment (phoinixgrr) [reply .circleci/config.yml:345]: Would you be ok with changing that to: 
```
 EXCLUDE_ENTERPRISE=TRUE make setup-go-work 
 make templates-archive
```

Or you need the `make templates-archive` step moved to a different stage? 
Let me know.

**Comment by agnivade** [comment]
> As I understand, This failure is not part of the changes introduced in this PR.

> Should we force-merge this and raise another ticket to troubleshoot the failures in enterprise ?

I see. So this seems like a flaky failure. In that case, might we push another dummy commit to see if it passes? IMO, all stages of CI should be green before merging a commit. But if you see a clear failure which is flaky, then no objections from my side in force-merging.

File: .circleci/config.yml (hunk 2 above)
Comment (phoinixgrr) [reply .circleci/config.yml:345]: Thanks for the comment @wiggin77  

I see that the Makefile is [defined by default](https://github.com/mattermost/focalboard/blob/main/Makefile#L42)  in the Makefile . Therefore setting `EXCLUDE_ENTERPRISE=true` in my proposal  can actually be omitted
(Just hope that the upstream makefile does not change this behaviour in the future, cause if it does it will brake this build. Therefore it is a good practise defining this). 

Concerning the argument of exporting the Environment variable `EXCLUDE_ENTERPRISE` on the  `templates-archive` target, instead of the `setup-go-work` , being ignored, that's not actually the case. 

Example checkingout Focalboard repository
```
❯ rm go.work
❯ EXCLUDE_ENTERPRISE=FALSE  make templates-archive
go run ./build/gowork/main.go
go.work written successfully.
go 1.18

use ./server
use ../mattermost-server
use ../enterprise
use ./mattermost-plugin

cd server/assets/build-template-archive; go run -tags 'json1' main.go --dir="../templates-boardarchive" --out="../templates.boardarchive"
```

It is clearly shown, that running the `make templates-archive` target with env variable set to: `EXCLUDE_ENTERPRISE=FALSE` , `setup-go-work` that is being called automatically as dependancy, it clearly automatically sets correctly the variable (EXCLUDE_ENTERPRISE is not ignored) , as it has printed `use ../enterprise` 

---
To sum this up. 

Would you be ok with changing this to 
```
 make templates-archive
 ``` 

Just to remind, the goal is to: 
1. Setup the go.work file without enterprise repo
2. run  `templates-archive` process

**Comment by agnivade** [comment]
Hmm .. it does not appear to be a flaky failure. @phoinixgrr - One thing to check would be the imports/imports.go file. That is copied over from the enterprise/imports/imports.go to mattermost-server/imports/imports.go. Can we check if that's happening?

Usually that is done in this part of the Makefile:

```
ifeq ($(BUILD_ENTERPRISE_READY),true)
	IGNORE:=$(shell echo Enterprise build selected, preparing)
	IGNORE:=$(shell rm -f imports/imports.go)
	IGNORE:=$(shell cp $(BUILD_ENTERPRISE_DIR)/imports/imports.go imports/)
	IGNORE:=$(shell rm -f enterprise)
else
```

**Comment by phoinixgrr** [comment]
> 

@agnivade  
Since this seems to be a problem introduced with the multi-product architecture changes, and related to golang-cli failures : 
```
WARN [runner] The linter 'golint' is deprecated (since v1.41.0) due to: The repository of the linter has been archived by the owner.  Replaced by revive. 
WARN [runner] Can't run linter goanalysis_metalinter: buildir: failed to load package cluster: could not load export data: no export data for "github.com/mattermost/enterprise/cluster" 
WARN [linters context] structcheck is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649. 
ERRO Running error: 1 error occurred:
	* can't run linter goanalysis_metalinter: buildir: failed to load package cluster: could not load export data: no export data for "github.com/mattermost/enterprise/cluster"
 
 ```
 
We can provide an ssh session, with the CircleCI Instance that runs the actual test that fails. The actual workspace that performs the tests will be available and mounted on the instace. 
Would you be ok with providing that info for you to troubleshoot this?

**Comment by agnivade** [comment]
Yep, that works

File: .circleci/config.yml (hunk 2 above)
Comment (phoinixgrr) [reply .circleci/config.yml:345]: @wiggin77  Any news on this?

**Review by pfltdv — APPROVED** [review]
:100: Great work Akis.

**Review by isacikgoz — APPROVED** [review]
Good to go after updating the https://github.com/mattermost/mattermost-server/tree/feature/CLD-3801-apply-multiproduct-arch with origin/master

**Review by spirosoik — APPROVED** [review]

**Review by wiggin77 — APPROVED** [review]
